services/src/main/java/org/keycloak/services/resources/admin/GroupsResource.java

```diff
@@ -77,17 +78,37 @@ public GroupsResource(RealmModel realm, KeycloakSession session, AdminPermission
                                                @QueryParam("max") Integer maxResults) {
         auth.groups().requireList();
 
-        List<GroupRepresentation> results;
+        List<GroupModel> results;
 
+        /**
+         * For now fetching (as in UsersResource) all groups and filtering in Java
+         * TODO: IMHO we should consider different pagination technique:
+         * (1) REST return filtered results (in some cases even epty page) and JS
+         * client retrieves next pages until it gets all page of results or reaches the end of list
+         * (2) or we fetch results in backend page by page until we get all required results.
+         */
         if (Objects.nonNull(search)) {
-            results = ModelToRepresentation.searchForGroupByName(realm, search.trim(), firstResult, maxResults);
-        } else if(Objects.nonNull(firstResult) && Objects.nonNull(maxResults)) {
-            results = ModelToRepresentation.toGroupHierarchy(realm, false, firstResult, maxResults);
+            results = realm.searchForGroupByName(search.trim(),null, null);
```

Comment:
Shouldn't be here "firstResult, maxResults" instead of "null, null" ?

Reply:
Indeed. @br0xpl, could you consider both `firstResult` and `maxResults` when searching by name ?

Reply:
The problem is we are getting maxResults then and some of the groups might get filtered out. So we would have to:
a) fetch maxResults in a loop until we get maxResults after filtering
b) fetch all 
I chose b - but if you prefer a just let me know?

@pedroigor , @mposolda : let me know how would you see it done?

But indeed my code misses the filtering first and max in after filtering - this has to be done.
- [ ] Filter first and max results out of the filtered results

Reply:
@br0xpl, another option would be:

* Check if the user has permissions to view groups (granted by view-users or manage-users)
* If not, obtain the groups that have permission enabled and return those that the user can view
* You can then use the list of groups to filter the results when searching by name

Look at this code in users https://github.com/keycloak/keycloak/blob/b4b3527df7656bfaaee351ec414e56c683c134c2/services/src/main/java/org/keycloak/services/resources/admin/UsersResource.java#L239. We set the list of groups that the user has view permission in the `KeycloakSession`. Then we use this list to filter results in the database.

You should also consider filtering top level groups. Just like it is done today.

Reply:
@br0xpl, see https://github.com/pedroigor/keycloak/tree/KEYCLOAK-10695. Similar to users. But still missing the top-level group pagination. Just a quick work to give you some ideas :)

Reply:
@pedroigor , @mposolda : I've fixed it according to your description - please see it. 

I've used the getGroupsWithViewPermission function of GroupPermissions but I have a doubt about it's behaviour, see here:

https://github.com/keycloak/keycloak/blob/b4b3527df7656bfaaee351ec414e56c683c134c2/services/src/main/java/org/keycloak/services/resources/admin/permissions/GroupPermissions.java#L306-L329

and line 323 - it is checking if user has manage**_members** or view**_members** permission. Shouldn't it be checking if user has view or manage permission for that group? The function name suggest just "view", not view_members..? 

Seems we need to either change function name and create another function with view or we need to change this function?

Reply:
We can change the method name. Its intent is to obtain the groups that a user has either view or manage members permission. Good catch.

Reply:
I've done that. I moved the common part to generic getGroupsWithPermissions(String... scopes) function and created two functions: getGroupsWithViewPermission /new function/, and renamed the previous: getGroupsWithViewMembersPermission. 

@pedroigor : please review all the changes -> I've tested and the groups work as expected now.